File: python3/search_in_rotated_sorted_array_II.py
"""
Space   : O(1)
Time    : O(log n)
"""


class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        if len(nums) == 0:
            return False

        s, e = 0, len(nums)-1

        while s <= e:
            mid = (s+e) // 2
            if nums[mid] == target:
                return True

            # check left
            if nums[s] < nums[mid]:
                if nums[s] <= target < nums[mid]:
                    e = mid - 1
                else:
                    s = mid + 1
            elif nums[s] == nums[mid]:
                s += 1
            # check right
            else:
                if nums[mid] < target <= nums[e]:
                    s = mid + 1
                else:
                    e = mid - 1

        return False

File: golang/search_in_rotated_sorted_array_ii.go
package main

/*
Space	: O(1)
Time	: O(log n)
*/

func searchRotatedArrayII(nums []int, target int) bool {
	n := len(nums)
	left, right := 0, n-1

	for left <= right {
		mid := (left + right) / 2
		if nums[mid] == target {
			return true
		}

		if nums[left] < nums[mid] {
			if nums[left] <= target && target < nums[mid] {
				right = mid - 1
			} else {
				left = mid + 1
			}
		} else if nums[left] == nums[mid] {
			left++
		} else {
			if nums[mid] < target && target <= nums[right] {
				left = mid + 1
			} else {
				right = mid - 1
			}
		}
	}

	return false
}
